File: weather/providers/us_nws.py
"""
United States National Weather Service (NWS) Weather Provider.

Fetches weather data from NWS public API.
Free, no API key required. USA coverage only.

API Documentation: https://www.weather.gov/documentation/services-web-api
"""

import asyncio
import re
from datetime import timezone, datetime, date
from typing import Optional
import urllib.request
import json

from .base import WeatherProvider, ProviderError, LocationNotSupportedError
from ..models import WeatherData, WeatherCondition, Location

# NWS API endpoints
NWS_BASE_URL = "https://api.weather.gov"
NWS_POINTS_URL = f"{NWS_BASE_URL}/points/{{lat}},{{lon}}"
NWS_STATIONS_URL = f"{NWS_BASE_URL}/stations"

# US city coordinates (latitude, longitude)
US_CITIES = {
    # Major metros
    "new york": (40.7128, -74.0060),
    "nyc": (40.7128, -74.0060),
    "los angeles": (34.0522, -118.2437),
    "la": (34.0522, -118.2437),
    "chicago": (41.8781, -87.6298),
    "houston": (29.7604, -95.3698),
    "phoenix": (33.4484, -112.0740),
    "philadelphia": (39.9526, -75.1652),
    "philly": (39.9526, -75.1652),
    "san antonio": (29.4241, -98.4936),
    "san diego": (32.7157, -117.1611),
    "dallas": (32.7767, -96.7970),
    "san jose": (37.3382, -121.8863),
    "austin": (30.2672, -97.7431),
    "jacksonville": (30.3322, -81.6557),
    "fort worth": (32.7555, -97.3308),
    "columbus": (39.9612, -82.9988),
    "charlotte": (35.2271, -80.8431),
    "san francisco": (37.7749, -122.4194),
    "sf": (37.7749, -122.4194),
    "indianapolis": (39.7684, -86.1581),
    "seattle": (47.6062, -122.3321),
    "denver": (39.7392, -104.9903),
    "washington dc": (38.9072, -77.0369),
    "dc": (38.9072, -77.0369),
    "boston": (42.3601, -71.0589),
    "nashville": (36.1627, -86.7816),
    "detroit": (42.3314, -83.0458),
    "portland": (45.5152, -122.6784),
    "las vegas": (36.1699, -115.1398),
    "memphis": (35.1495, -90.0490),
    "louisville": (38.2527, -85.7585),
    "baltimore": (39.2904, -76.6122),
    "milwaukee": (43.0389, -87.9065),
    "albuquerque": (35.0844, -106.6504),
    "tucson": (32.2226, -110.9747),
    "fresno": (36.7378, -119.7871),
    "sacramento": (38.5816, -121.4944),
    "kansas city": (39.0997, -94.5783),
    "atlanta": (33.7490, -84.3880),
    "miami": (25.7617, -80.1918),
    "orlando": (28.5383, -81.3792),
    "minneapolis": (44.9778, -93.2650),
    "pittsburgh": (40.4406, -79.9959),
    "st louis": (38.6270, -90.1994),
    "cleveland": (41.4993, -81.6944),
    "new orleans": (29.9511, -90.0715),
    "tampa": (27.9506, -82.4572),
    "honolulu": (21.3069, -157.8583),
    "anchorage": (61.2181, -149.9003),
}

# NWS weather condition mapping
NWS_CONDITION_MAP = {
    # Clear/Sunny
    "skc": WeatherCondition.CLEAR,       # Sky Clear
    "few": WeatherCondition.SUNNY,       # Few Clouds
    "sct": WeatherCondition.PARTLY_CLOUDY,  # Scattered Clouds
    "bkn": WeatherCondition.CLOUDY,      # Broken Clouds
    "ovc": WeatherCondition.OVERCAST,    # Overcast

    # Rain
    "ra": WeatherCondition.RAIN,         # Rain
    "-ra": WeatherCondition.DRIZZLE,     # Light Rain
    "+ra": WeatherCondition.HEAVY_RAIN,  # Heavy Rain
    "rw": WeatherCondition.SHOWERS,      # Rain Showers
    "-rw": WeatherCondition.SHOWERS,     # Light Rain Showers
    "+rw": WeatherCondition.HEAVY_RAIN,  # Heavy Rain Showers
    "ts": WeatherCondition.THUNDERSTORM,  # Thunderstorm
    "-ts": WeatherCondition.THUNDERSTORM, # Light Thunderstorm
    "+ts": WeatherCondition.THUNDERSTORM, # Heavy Thunderstorm

    # Snow
    "sn": WeatherCondition.SNOW,         # Snow
    "-sn": WeatherCondition.SNOW,        # Light Snow
    "+sn": WeatherCondition.HEAVY_SNOW,  # Heavy Snow
    "sw": WeatherCondition.SNOW,         # Snow Showers
    "-sw": WeatherCondition.SNOW,        # Light Snow Showers
    "+sw": WeatherCondition.HEAVY_SNOW,  # Heavy Snow Showers

    # Mixed
    "ip": WeatherCondition.SLEET,        # Ice Pellets
    "fzra": WeatherCondition.SLEET,      # Freezing Rain
    "zyr": WeatherCondition.SLEET,       # Freezing Drizzle

    # Fog/Mist/Haze
    "fg": WeatherCondition.FOG,          # Fog
    "br": WeatherCondition.MIST,         # Mist
    "hz": WeatherCondition.MIST,         # Haze
    "fu": WeatherCondition.FOG,          # Smoke

    # Wind
    "du": WeatherCondition.WINDY,        # Dust
    "sa": WeatherCondition.WINDY,        # Sand
    "po": WeatherCondition.WINDY,        # Volcanic Ash
}

# Supported locations
SUPPORTED_LOCATIONS = {
    "usa", "united states", "america", "us",
    "new york", "los angeles", "chicago", "houston",
    "phoenix", "philadelphia", "san antonio", "san diego",
    "dallas", "san jose", "austin", "jacksonville",
    "fort worth", "columbus", "charlotte", "san francisco",
    "indianapolis", "seattle", "denver", "washington dc",
    "boston", "nashville", "detroit", "portland",
    "las vegas", "memphis", "louisville", "baltimore",
    "milwaukee", "albuquerque", "tucson", "fresno",
    "sacramento", "kansas city", "atlanta", "miami",
    "orlando", "minneapolis", "pittsburgh", "st louis",
    "cleveland", "new orleans", "tampa", "honolulu", "anchorage",
}


class NWSProvider(WeatherProvider):
    """
    US National Weather Service weather provider.

    - Free, no API key required
    - USA coverage only
    - Provides current weather and 7-day forecast
    """

    priority = 7  # After other regional providers
    supports_forecast = True
    supports_air_quality = False
    requires_api_key = False

    @property
    def name(self) -> str:
        return "nws"

    def supports_location(self, location: Location) -> bool:
        """Check if location is in the USA."""
        normalized = location.normalized.lower()
        return normalized in SUPPORTED_LOCATIONS or any(
            city in normalized for city in US_CITIES
        )

    async def get_current(self, location: Location) -> WeatherData:
        """Fetch current weather for US location."""
        if not self.supports_location(location):
            raise LocationNotSupportedError(
                f"NWS only supports US locations: {location.raw}"
            )

        try:
            lat, lon = self._get_coordinates(location)
            grid_data = await self._fetch_gridpoint(lat, lon)
            station_data = await self._fetch_observations(grid_data)
            return self._parse_current(location, station_data)
        except Exception as e:
            raise ProviderError(f"NWS API error: {e}")

    async def get_forecast(
        self,
        location: Location,
        days: int = 7
    ) -> list[WeatherData]:
        """Fetch weather forecast for US location."""
        if not self.supports_location(location):
            raise LocationNotSupportedError(
                f"NWS only supports US locations: {location.raw}"
            )

        try:
            lat, lon = self._get_coordinates(location)
            grid_data = await self._fetch_gridpoint(lat, lon)
            forecast_data = await self._fetch_forecast(grid_data)
            return self._parse_forecast(location, forecast_data, days)
        except Exception as e:
            raise ProviderError(f"NWS API error: {e}")

    def _get_coordinates(self, location: Location) -> tuple[float, float]:
        """Get coordinates for US location."""
        normalized = location.normalized.lower()

        for city, coords in US_CITIES.items():
            if city in normalized:
                return coords

        # Default to New York
        return US_CITIES["new york"]

    async def _fetch_gridpoint(self, latitude: float, longitude: float) -> dict:
        """Fetch gridpoint data from NWS to get forecast URLs."""
        url = NWS_POINTS_URL.format(lat=latitude, lon=longitude)

        loop = asyncio.get_event_loop()

        def fetch():
            req = urllib.request.Request(url)
            req.add_header("User-Agent", "WeatherSkill/1.0 (support@weather-skill.io)")
            req.add_header("Accept", "application/json")
            with urllib.request.urlopen(req, timeout=10) as response:
                return json.loads(response.read().decode('utf-8'))

        return await loop.run_in_executor(None, fetch)

    async def _fetch_observations(self, grid_data: dict) -> dict:
        """Fetch current observations from nearest station."""
        # Get observation stations URL
        stations_url = grid_data.get("properties", {}).get("observationStations")

        if not stations_url:
            raise ProviderError("No observation stations available")

        loop = asyncio.get_event_loop()

        def fetch():
            req = urllib.request.Request(stations_url)
            req.add_header("User-Agent", "WeatherSkill/1.0 (support@weather-skill.io)")
            req.add_header("Accept", "application/json")
            with urllib.request.urlopen(req, timeout=10) as response:
                stations_data = json.loads(response.read().decode('utf-8'))

            # Get first station's observations
            station_id = stations_data.get("features", [{}])[0].get("properties", {}).get("stationIdentifier")
            if not station_id:
                raise ProviderError("No station identifier found")

            obs_url = f"{NWS_BASE_URL}/stations/{station_id}/observations/latest"
            req = urllib.request.Request(obs_url)
            req.add_header("User-Agent", "WeatherSkill/1.0 (support@weather-skill.io)")
            req.add_header("Accept", "application/json")
            with urllib.request.urlopen(req, timeout=10) as response:
                return json.loads(response.read().decode('utf-8'))

        return await loop.run_in_executor(None, fetch)

    async def _fetch_forecast(self, grid_data: dict) -> dict:
        """Fetch forecast data from NWS."""
        forecast_url = grid_data.get("properties", {}).get("forecast")

        if not forecast_url:
            raise ProviderError("No forecast URL available")

        loop = asyncio.get_event_loop()

        def fetch():
            req = urllib.request.Request(forecast_url)
            req.add_header("User-Agent", "WeatherSkill/1.0 (support@weather-skill.io)")
            req.add_header("Accept", "application/json")
            with urllib.request.urlopen(req, timeout=10) as response:
                return json.loads(response.read().decode('utf-8'))

        return await loop.run_in_executor(None, fetch)

    def _parse_current(self, location: Location, data: dict) -> WeatherData:
        """Parse current weather from NWS response."""
        properties = data.get("properties", {})

        # Temperature (convert from Celsius to ensure consistency)
        temp_c = properties.get("temperature", {}).get("value")
        if temp_c:
            temp = temp_c
        else:
            temp = 0.0

        # Other observations
        humidity = properties.get("relativeHumidity", {}).get("value")
        wind_speed_mps = properties.get("windSpeed", {}).get("value")
        wind_speed = wind_speed_mps * 3.6 if wind_speed_mps else None  # m/s to km/h
        wind_dir = properties.get("windDirection", {}).get("value")
        pressure = properties.get("barometricPressure", {}).get("value")
        visibility = properties.get("visibility", {}).get("value")
        feels_like_c = properties.get("heatIndex", {}).get("value") or properties.get("windChill", {}).get("value")

        # Get text description
        description = properties.get("textDescription", "")

        # Map condition
        condition = self._map_condition(description)

        # Parse timestamp
        timestamp = properties.get("timestamp", "")
        try:
            observed_at = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
        except (ValueError, TypeError):
            observed_at = datetime.now(timezone.utc)

        # Get display name
        display_name = self._get_display_name(location)

        # Wind direction
        wind_direction = self._deg_to_direction(wind_dir) if wind_dir else None

        return WeatherData(
            location=display_name,
            temperature=temp if temp is not None else 0.0,
            feels_like=feels_like_c,
            humidity=int(humidity) if humidity else None,
            wind_speed=wind_speed,
            wind_direction=wind_direction,
            pressure=pressure / 100 if pressure else None,  # Pa to hPa
            visibility=visibility / 1000 if visibility else None,  # m to km
            condition=condition,
            condition_raw=description,
            description=description,
            observed_at=observed_at,
            fetched_at=datetime.now(timezone.utc),
            provider_name=self.name,
        )

    def _parse_forecast(
        self,
        location: Location,
        data: dict,
        days: int
    ) -> list[WeatherData]:
        """Parse forecast from NWS response."""
        properties = data.get("properties", {})
        periods = properties.get("periods", [])

        results = []
        display_name = self._get_display_name(location)
        seen_dates = set()

        for period in periods:
            if len(results) >= days:
                break

            # Get date
            start_time = period.get("startTime", "")
            try:
                fc_datetime = datetime.fromisoformat(start_time.replace("Z", "+00:00"))
                fc_date = fc_datetime.date()
            except (ValueError, TypeError):
                continue

            # Skip if we already have this date (NWS returns day and night periods)
            if fc_date in seen_dates:
                continue
            seen_dates.add(fc_date)

            # Get temperatures
            temp = period.get("temperature", 0)
            temp_unit = period.get("temperatureUnit", "F")

            # Convert to Celsius if needed
            if temp_unit == "F":
                temp_c = (temp - 32) * 5 / 9
            else:
                temp_c = temp

            # Get condition
            short_forecast = period.get("shortForecast", "")
            condition = self._map_condition(short_forecast)

            # Wind
            wind_str = period.get("windSpeed", "")
            wind_speed = self._parse_wind_speed(wind_str)
            wind_dir = period.get("windDirection", "")

            results.append(WeatherData(
                location=display_name,
                temperature=temp_c,
                temp_high=temp_c if period.get("isDaytime") else None,
                temp_low=temp_c if not period.get("isDaytime") else None,
                forecast_date=fc_date,
                condition=condition,
                condition_raw=short_forecast,
                description=period.get("detailedForecast", ""),
                wind_speed=wind_speed,
                wind_direction=wind_dir,
                fetched_at=datetime.now(timezone.utc),
                provider_name=self.name,
            ))

        return results

    def _get_display_name(self, location: Location) -> str:
        """Get display name for location."""
        normalized = location.normalized.lower()

        for city in US_CITIES:
            if city in normalized:
                return city.title()

        return "United States"

    def _map_condition(self, condition_str: str) -> WeatherCondition:
        """Map NWS condition text to WeatherCondition."""
        if not condition_str:
            return WeatherCondition.UNKNOWN

        normalized = condition_str.lower().strip()

        # Keyword matching (NWS uses descriptive text)
        if "sunny" in normalized or "clear" in normalized or "fair" in normalized:
            return WeatherCondition.SUNNY
        if "partly cloud" in normalized or "mostly sunny" in normalized:
            return WeatherCondition.PARTLY_CLOUDY
        if "cloud" in normalized or "overcast" in normalized:
            return WeatherCondition.CLOUDY
        if "shower" in normalized:
            return WeatherCondition.SHOWERS
        if "rain" in normalized:
            return WeatherCondition.RAIN
        if "thunder" in normalized or "t-storm" in normalized:
            return WeatherCondition.THUNDERSTORM
        if "snow" in normalized:
            return WeatherCondition.SNOW
        if "fog" in normalized:
            return WeatherCondition.FOG
        if "wind" in normalized or "breezy" in normalized:
            return WeatherCondition.WINDY
        if "sleet" in normalized or "freezing rain" in normalized:
            return WeatherCondition.SLEET
        if "hail" in normalized:
            return WeatherCondition.HAIL

        return WeatherCondition.UNKNOWN

    def _parse_wind_speed(self, wind_str: str) -> Optional[float]:
        """Parse wind speed from NWS string like '10 to 15 mph'."""
        if not wind_str:
            return None

        # Extract numbers from string like "10 to 15 mph" or "10 mph"
        numbers = re.findall(r'\d+', wind_str)
        if numbers:
            # Take average if range, or single value
            speeds = [int(n) for n in numbers]
            avg_mph = sum(speeds) / len(speeds)
            return avg_mph * 1.60934  # mph to km/h
        return None

    def _deg_to_direction(self, deg: Optional[float]) -> Optional[str]:
        """Convert wind degrees to cardinal direction."""
        if deg is None:
            return None

        directions = ["N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
                      "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"]
        index = round(deg / 22.5) % 16
        return directions[index]

File: src/providers/us_nws.ts
/* eslint-disable @typescript-eslint/naming-convention */
/**
 * United States National Weather Service (NWS) Weather Provider.
 *
 * Fetches weather data from NWS public API.
 * Free, no API key required. USA coverage only.
 *
 * API Documentation: https://www.weather.gov/documentation/services-web-api
 *
 * Mirrors `weather/providers/us_nws.py`.
 */

import { US_NWS_CITIES } from "../data-loader.js";
import { makeWeatherData } from "../models.js";
import type { IWeatherProvider, Location, WeatherData } from "../types.js";
import {
  LocationNotSupportedError,
  ProviderError,
  WeatherCondition,
} from "../types.js";

// NWS API endpoints
const NWS_BASE_URL = "https://api.weather.gov";

// Required by NWS — requests without this header are rejected with 403.
// Mirrors `weather/providers/us_nws.py:133`.
const NWS_USER_AGENT = "WeatherSkill/1.0 (support@weather-skill.io)";

// Supported locations (mirror weather/providers/us_nws.py:39-52)
const SUPPORTED_LOCATIONS: ReadonlySet<string> = new Set([
  "usa", "united states", "america", "us",
  "new york", "los angeles", "chicago", "houston",
  "phoenix", "philadelphia", "san antonio", "san diego",
  "dallas", "san jose", "austin", "jacksonville",
  "fort worth", "columbus", "charlotte", "san francisco",
  "indianapolis", "seattle", "denver", "washington dc",
  "boston", "nashville", "detroit", "portland",
  "las vegas", "memphis", "louisville", "baltimore",
  "milwaukee", "albuquerque", "tucson", "fresno",
  "sacramento", "kansas city", "atlanta", "miami",
  "orlando", "minneapolis", "pittsburgh", "st louis",
  "cleveland", "new orleans", "tampa", "honolulu", "anchorage",
]);

const CARDINAL_DIRECTIONS = [
  "N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
  "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW",
] as const;

/**
 * US National Weather Service weather provider.
 *
 * - Free, no API key required
 * - USA coverage only
 * - Provides current weather and 7-day forecast
 */
export class NWSProvider implements IWeatherProvider {
  readonly name = "nws";
  readonly priority = 7; // After other regional providers
  readonly supportsForecast = true;
  readonly supportsAirQuality = false;
  readonly requiresApiKey = false;

  /** Check if location is in the USA. */
  supportsLocation(location: Location): boolean {
    const normalized = location.normalized.toLowerCase();
    if (SUPPORTED_LOCATIONS.has(normalized)) return true;
    for (const city of Object.keys(US_NWS_CITIES)) {
      if (normalized.includes(city)) return true;
    }
    return false;
  }

  /** Fetch current weather for US location. */
  async getCurrent(location: Location): Promise<WeatherData> {
    if (!this.supportsLocation(location)) {
      throw new LocationNotSupportedError(
        `NWS only supports US locations: ${location.raw}`,
      );
    }

    try {
      const [lat, lon] = this.getCoordinates(location);
      const gridData = await this.fetchGridpoint(lat, lon);
      const stationData = await this.fetchObservations(gridData);
      return this.parseCurrent(location, stationData);
    } catch (e) {
      if (e instanceof LocationNotSupportedError) throw e;
      const msg = e instanceof Error ? e.message : String(e);
      throw new ProviderError(`NWS API error: ${msg}`);
    }
  }

  /** Fetch weather forecast for US location. */
  async getForecast(
    location: Location,
    days: number = 7,
  ): Promise<WeatherData[]> {
    if (!this.supportsLocation(location)) {
      throw new LocationNotSupportedError(
        `NWS only supports US locations: ${location.raw}`,
      );
    }

    try {
      const [lat, lon] = this.getCoordinates(location);
      const gridData = await this.fetchGridpoint(lat, lon);
      const forecastData = await this.fetchForecast(gridData);
      return this.parseForecast(location, forecastData, days);
    } catch (e) {
      if (e instanceof LocationNotSupportedError) throw e;
      const msg = e instanceof Error ? e.message : String(e);
      throw new ProviderError(`NWS API error: ${msg}`);
    }
  }

  /** Get coordinates for US location. Defaults to New York. */
  private getCoordinates(location: Location): [number, number] {
    const normalized = location.normalized.toLowerCase();
    for (const [city, coords] of Object.entries(US_NWS_CITIES)) {
      if (normalized.includes(city)) return coords;
    }
    return US_NWS_CITIES["new york"]!;
  }

  /**
   * GET an NWS endpoint with the required `User-Agent` header. Without
   * this header NWS returns HTTP 403.
   */
  private async fetchJson(url: string): Promise<Record<string, any>> {
    const response = await fetch(url, {
      headers: {
        "User-Agent": NWS_USER_AGENT,
        "Accept": "application/json",
      },
    });
    if (!response.ok) {
      throw new Error(`HTTP ${response.status} ${response.statusText}`);
    }
    return (await response.json()) as Record<string, any>;
  }

  /** Fetch gridpoint metadata (forecast/station URLs) for coordinates. */
  private async fetchGridpoint(
    latitude: number,
    longitude: number,
  ): Promise<Record<string, any>> {
    const url = `${NWS_BASE_URL}/points/${latitude},${longitude}`;
    return this.fetchJson(url);
  }

  /** Fetch current observations from nearest station. */
  private async fetchObservations(
    gridData: Record<string, any>,
  ): Promise<Record<string, any>> {
    const stationsUrl: string | undefined =
      gridData.properties?.observationStations;
    if (!stationsUrl) {
      throw new ProviderError("No observation stations available");
    }

    const stationsData = await this.fetchJson(stationsUrl);
    const features: Record<string, any>[] = stationsData.features ?? [];
    const stationId: string | undefined =
      features[0]?.properties?.stationIdentifier;
    if (!stationId) {
      throw new ProviderError("No station identifier found");
    }

    const obsUrl = `${NWS_BASE_URL}/stations/${stationId}/observations/latest`;
    return this.fetchJson(obsUrl);
  }

  /** Fetch forecast data from NWS. */
  private async fetchForecast(
    gridData: Record<string, any>,
  ): Promise<Record<string, any>> {
    const forecastUrl: string | undefined = gridData.properties?.forecast;
    if (!forecastUrl) {
      throw new ProviderError("No forecast URL available");
    }
    return this.fetchJson(forecastUrl);
  }

  /** Parse current weather from NWS observation response. */
  private parseCurrent(
    location: Location,
    data: Record<string, any>,
  ): WeatherData {
    const properties: Record<string, any> = data.properties ?? {};

    // Temperature (NWS observation values are already Celsius)
    const tempC: number | null | undefined = properties.temperature?.value;
    const temp = tempC != null ? tempC : 0.0;

    // Other observations
    const humidity: number | null | undefined =
      properties.relativeHumidity?.value;
    const windSpeedMps: number | null | undefined = properties.windSpeed?.value;
    const windSpeed = windSpeedMps != null ? windSpeedMps * 3.6 : undefined; // m/s → km/h
    const windDir: number | null | undefined = properties.windDirection?.value;
    const pressure: number | null | undefined =
      properties.barometricPressure?.value;
    const visibility: number | null | undefined = properties.visibility?.value;
    const feelsLikeC: number | null | undefined =
      properties.heatIndex?.value ?? properties.windChill?.value;

    // Text description
    const description: string = properties.textDescription ?? "";
    const condition = this.mapCondition(description);

    // Parse timestamp (ISO-8601, may end in `Z`)
    const timestamp: string = properties.timestamp ?? "";
    let observedAt: Date;
    const parsed = timestamp ? new Date(timestamp) : new Date(NaN);
    if (!Number.isNaN(parsed.getTime())) {
      observedAt = parsed;
    } else {
      observedAt = new Date();
    }

    const displayName = this.getDisplayName(location);
    const windDirection = windDir != null ? this.degToDirection(windDir) : undefined;

    return makeWeatherData({
      location: displayName,
      temperature: temp,
      ...(feelsLikeC != null ? { feels_like: feelsLikeC } : {}),
      ...(humidity != null ? { humidity: Math.trunc(humidity) } : {}),
      ...(windSpeed != null ? { wind_speed: windSpeed } : {}),
      ...(windDirection != null ? { wind_direction: windDirection } : {}),
      ...(pressure != null ? { pressure: pressure / 100 } : {}), // Pa → hPa
      ...(visibility != null ? { visibility: visibility / 1000 } : {}), // m → km
      condition,
      condition_raw: description,
      ...(description ? { description } : {}),
      observed_at: observedAt,
      fetched_at: new Date(),
      provider_name: this.name,
    });
  }

  /** Parse forecast from NWS response. */
  private parseForecast(
    location: Location,
    data: Record<string, any>,
    days: number,
  ): WeatherData[] {
    const properties: Record<string, any> = data.properties ?? {};
    const periods: Record<string, any>[] = properties.periods ?? [];

    const results: WeatherData[] = [];
    const displayName = this.getDisplayName(location);
    const seenDates = new Set<string>();

    for (const period of periods) {
      if (results.length >= days) break;

      const startTime: string = period.startTime ?? "";
      const fcDateTime = startTime ? new Date(startTime) : new Date(NaN);
      if (Number.isNaN(fcDateTime.getTime())) continue;

      // Mirror Python's `datetime.fromisoformat(...).date()` — preserve
      // the LOCAL calendar date from the ISO offset, not UTC. Otherwise
      // a US-eastern-time period like "2026-05-13T23:00:00-04:00"
      // (local: 2026-05-13) would dedupe as 2026-05-14 in UTC and
      // misassign the forecast day.
      const localDateMatch = startTime.match(/^(\d{4}-\d{2}-\d{2})/);
      if (!localDateMatch) continue;
      const dateKey = localDateMatch[1]!;
      if (seenDates.has(dateKey)) continue;
      seenDates.add(dateKey);

      // Forecast date (midnight UTC of the period's local date —
      // matches Python's naive `datetime.date()` JSON serialization).
      const fcDate = new Date(`${dateKey}T00:00:00Z`);

      // Temperature — NWS forecast uses Fahrenheit by default.
      const temp: number = period.temperature ?? 0;
      const tempUnit: string = period.temperatureUnit ?? "F";
      const tempC = tempUnit === "F" ? ((temp - 32) * 5) / 9 : temp;

      const shortForecast: string = period.shortForecast ?? "";
      const condition = this.mapCondition(shortForecast);

      const windStr: string = period.windSpeed ?? "";
      const windSpeed = this.parseWindSpeed(windStr);
      const windDir: string = period.windDirection ?? "";
      const detailedForecast: string = period.detailedForecast ?? "";
      const isDaytime: boolean = period.isDaytime === true;

      results.push(
        makeWeatherData({
          location: displayName,
          temperature: tempC,
          ...(isDaytime ? { temp_high: tempC } : { temp_low: tempC }),
          forecast_date: fcDate,
          condition,
          condition_raw: shortForecast,
          ...(detailedForecast ? { description: detailedForecast } : {}),
          ...(windSpeed != null ? { wind_speed: windSpeed } : {}),
          ...(windDir ? { wind_direction: windDir } : {}),
          fetched_at: new Date(),
          provider_name: this.name,
        }),
      );
    }

    return results;
  }

  /** Get display name for location. */
  private getDisplayName(location: Location): string {
    const normalized = location.normalized.toLowerCase();
    for (const city of Object.keys(US_NWS_CITIES)) {
      if (normalized.includes(city)) {
        return city.replace(/\b\w/g, (c) => c.toUpperCase());
      }
    }
    return "United States";
  }

  /**
   * Map NWS condition text to WeatherCondition via keyword matching.
   * Mirrors `_map_condition` in `weather/providers/us_nws.py`.
   *
   * Note: We intentionally do NOT use `NWS_CONDITION_MAP` from the
   * data loader — that map is keyed on METAR codes (`skc`, `few`,
   * `bkn`, …), not the descriptive text NWS returns in
   * `textDescription` / `shortForecast`.
   */
  private mapCondition(conditionStr: string): WeatherCondition {
    if (!conditionStr) return WeatherCondition.Unknown;
    const normalized = conditionStr.toLowerCase().trim();

    if (
      normalized.includes("sunny") ||
      normalized.includes("clear") ||
      normalized.includes("fair")
    ) return WeatherCondition.Sunny;
    if (
      normalized.includes("partly cloud") ||
      normalized.includes("mostly sunny")
    ) return WeatherCondition.PartlyCloudy;
    if (normalized.includes("cloud") || normalized.includes("overcast")) {
      return WeatherCondition.Cloudy;
    }
    if (normalized.includes("shower")) return WeatherCondition.Showers;
    if (normalized.includes("rain")) return WeatherCondition.Rain;
    if (normalized.includes("thunder") || normalized.includes("t-storm")) {
      return WeatherCondition.Thunderstorm;
    }
    if (normalized.includes("snow")) return WeatherCondition.Snow;
    if (normalized.includes("fog")) return WeatherCondition.Fog;
    if (normalized.includes("wind") || normalized.includes("breezy")) {
      return WeatherCondition.Windy;
    }
    if (normalized.includes("sleet") || normalized.includes("freezing rain")) {
      return WeatherCondition.Sleet;
    }
    if (normalized.includes("hail")) return WeatherCondition.Hail;

    return WeatherCondition.Unknown;
  }

  /**
   * Parse wind speed from NWS string like `"10 to 15 mph"` or
   * `"10 mph"`. Returns km/h, or `undefined` if no number is found.
   */
  private parseWindSpeed(windStr: string): number | undefined {
    if (!windStr) return undefined;
    const matches = windStr.match(/\d+/g);
    if (!matches || matches.length === 0) return undefined;
    const speeds = matches.map((n) => Number.parseInt(n, 10));
    const avgMph = speeds.reduce((a, b) => a + b, 0) / speeds.length;
    return avgMph * 1.60934; // mph → km/h
  }

  /** Convert wind degrees to a 16-point cardinal direction string. */
  private degToDirection(deg: number): string {
    const index = Math.round(deg / 22.5) % 16;
    return CARDINAL_DIRECTIONS[(index + 16) % 16]!;
  }
}
